Refactor nodelocal-dns Helm chart into component

acumino opened the pull request:
Refactor

abdasgupta committed:
Added changes to support compaction job through etcd-druid.

acumino committed:
Refactor nodelocal-dns Helm chart into component

rfranzke committed:
Make remaining values optional for generic control plane actuator

Co-Authored-By: Tim Ebert <tim.ebert@sap.com>

rfranzke committed:
Allow object selector for extensions webhooks

Co-Authored-By: Tim Ebert <tim.ebert@sap.com>

istvanballok committed:
Capture a metric for a persistent volume issue (#5040)

We observed that sometimes persistent volumes are not mounted properly
and the local disk is used instead. To detect that situation, we compare
2 metrics:
- the size of the persistent volume claim, as exposed by kube-state-metrics
- the size of the mounted volume, as exposed by the kubelet

Due to the overhead of the filesystem, they are not exactly the same
(filesystem capacity is about 2% less than the raw disk capacity).

We expose these metrics for federation in the garden cluster, where an
alert could be defined to notify the operators.

Co-authored-by: Wesley Bermbach <wesley.bermbach@sap.com>
Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

Co-authored-by: Wesley Bermbach <wesley.bermbach@sap.com>

rfranzke committed:
Prepare generate-imagename-constants.sh script for reuse

Co-Authored-By: Tim Ebert <tim.ebert@sap.com>

stoyanr committed:
Move copying of seed conditions to shoots registered as seeds to GCM (#5013)

* Move updating conditions of shoots registered as seeds to GCM

* Add messages to ManagedSeed conditions

* Remove SyncPeriod and address other comments

* Fix handling of resource versions in ControlledResourceEventHandler

rfranzke committed:
Several enhancements for the integration test framework (#5050)

* InfrastructureConfig is optional

* Allow specifying networking type

* Drop dead code

* VolumeTypes are optional

* Do not enforce the shoot yaml to be in the templates directory

istvanballok committed:
Expose a metric on frequent container restarts in the seed (#5048)

This metric can be federated into the garden soil cluster to
emit alerts when a garden system component or extension is
restarting frequently.

Co-authored-by: Wesley Bermbach <wesley.bermbach@sap.com>
Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

Co-authored-by: Wesley Bermbach <wesley.bermbach@sap.com>

acumino committed:
Updated monitoring test

ialidzhikov commented on pkg/operation/botanist/botanist.go:
```diff
@@ -150,6 +150,16 @@ func New(ctx context.Context, o *operation.Operation) (*Botanist, error) {
 	}
+	//TODO(acumino): Need to add when node local dns is enabled
+	if o.Shoot.NodeLocalDNSEnabled {
+		o.Shoot.Components.SystemComponents.NodeLocalDNS, err = b.DefaultNodeLocalDNS()
+		if err != nil {
+			return nil, err
+		} else {
+			o.Shoot.Components.SystemComponents.NodeLocalDNS = nil
+		}
```
You can wrap ` b.DefaultNodeLocalDNS()` with `component.OpDestroy` if the node local dns is not enabled. Another option is to manually call Destroy of the NodeLocalDNS component in shoot_control_{reconcile,delete}.

ialidzhikov replied:
I cannot follow this `else` block. If there was not error constructing the DefaultNodeLocalDNS, then set it to nil? Why? Did you rather wanted to set the NodeLocalDNS field to nil when node local dns is not enabled?

ialidzhikov commented on charts/seed-monitoring/charts/core/charts/prometheus/templates/config.yaml:
```diff
@@ -307,40 +307,40 @@ data:
-      metric_relabel_configs:
-{{ include "prometheus.keep-metrics.metric-relabel-config" .Values.allowedMetrics.nodeLocalDNS | indent 6 }}
-{{ end }}
+# {{- if .Values.nodeLocalDNS.enabled }}
```
Most probably you commented this section (instead of removing it) to make it easier for development. But when you are about to create a PR to g/g, then you have to remove this section (and other commented sections as well).

ialidzhikov commented on charts/seed-monitoring/charts/core/charts/prometheus/values.yaml:
```diff
@@ -21,8 +21,8 @@ ingress:
 
-nodeLocalDNS:
-  enabled: false
+# nodeLocalDNS:
```
Same as above for the commented sections - they should be removed.

ialidzhikov commented on charts/seed-monitoring/charts/core/charts/prometheus/values.yaml:
```diff
@@ -53,21 +53,21 @@ allowedMetrics:
+  # - coredns_kubernetes_dns_programming_duration_seconds_count
+  # - coredns_kubernetes_dns_programming_duration_seconds_sum
+  # - process_max_fds
+  # - process_open_fds
```
Same as above for the commented sections - they should be removed.

ialidzhikov commented on pkg/operation/botanist/component/nodelocaldns/nodelocaldns.go:
```diff
@@ -0,0 +1,543 @@
+	prometheusPort   = 9253
+	prometheusScrape = true
+
+	nodeLocal = common.NodeLocalIPVSAddress     //"169.254.20.10"
```
Do we need this const? Can't we use directly `common.NodeLocalIPVSAddress` in the places where we need it?

ialidzhikov commented on pkg/operation/botanist/component/nodelocaldns/nodelocaldns.go:
```diff
@@ -0,0 +1,543 @@
+	prometheusScrape = true
+
+	nodeLocal = common.NodeLocalIPVSAddress     //"169.254.20.10"
+	domain    = gardencorev1beta1.DefaultDomain //"cluster.local"
```
Do we need this const? Can't we use directly `gardencorev1beta1.DefaultDomain` in the places where we need it?

ialidzhikov commented on pkg/operation/botanist/component/nodelocaldns/nodelocaldns.go:
```diff
@@ -0,0 +1,543 @@
+									},
+								},
+								Args: []string{
+									"-localip", //Todo add if else condition
```
Is this todo addressed or is it still relevant?

ialidzhikov commented on pkg/operation/care/health.go:
```diff
@@ -300,6 +301,8 @@ func init() {
 	utilruntime.Must(err)
 	versionConstraintGreaterEqual132, err = semver.NewConstraint(">= 1.32")
 	utilruntime.Must(err)
+	versionConstraintGreaterEqual136, err = semver.NewConstraint(">= 1.36")
```
It should be already 1.37 and Gardener 1.36 is already released.

ialidzhikov commented on pkg/operation/botanist/component/nodelocaldns/nodelocaldns.go:
```diff
@@ -0,0 +1,543 @@
+		}
+		vpa *autoscalingv1beta2.VerticalPodAutoscaler
+	)
+	utilruntime.Must(kutil.MakeUnique(configMap))
+	daemonset.Spec.Template.Spec.Volumes[2].VolumeSource.ConfigMap.LocalObjectReference.Name = configMap.Name
+	utilruntime.Must(references.InjectAnnotations(daemonset))
```
Can you rather move the configMap initialization and its MakeUnique call before the DaemonSet initialization? You can probably check how it is handled for the dependency-watchdog component.

ialidzhikov reviewed (commented):
Overall looks good. I am adding few inline comments from what I managed to spot.

stoyanr committed:
Add missing GRM network policy and network policy labels (#5052)

Commit:
Skip moving generated CRDs when prefix is empty (#5053)

stoyanr committed:
Move shoot status label updates to GCM (#5022)

plkokanov committed:
Modify restore operation to include copy of etcd backups (#4894)

* Vendor etcd

* Add support for source backupentries

* Modify restore operation to include copy of etcd backups

* Add unit tests

* Vendor etcd 0.7.0

* Add doc for CopyEtcdBackupsDuringControlPlaneMigration feature gate

* Address code review comments

* Read source provider from source backup entry

Co-authored-by: Stoyan Rachev <s.rachev@sap.com>

ary1992 committed:
Add integration test for ssh-key rotation (#4977)

* Add integration test for ssh-key rotation

Co-authored-by: Max Becker <max.becker@sap.com>

* Move test to operations package

Also now using constants instead of magic strings

Co-authored-by: Max Becker <max.becker@sap.com>

rfranzke committed:
Do not consider GRM webhooks in constraints check

rfranzke committed:
Eliminate client certificate for `kube-controller-manager` and `cluster-autoscaler` (#5007)

* Adapt `kube-controller-manager` to use shoot access token

* Adapt `cluster-autoscaler` to use shoot access token

rfranzke committed:
Adapt `kube-rbac-proxy` to use shoot access token (#5010)

On the way, I improved the data flow a bit to make it easier maintainable and more compact

timebertt committed:
Merge pull request #5042 from rfranzke/enh/ext-cp-optional

Make remaining values optional for generic control plane actuator

timebertt committed:
Merge pull request #5043 from rfranzke/enh/ext-wh-objectselector

Allow object selector for extensions webhooks

timebertt committed:
Merge pull request #5046 from rfranzke/enh/generate-imagename-constants

Prepare `generate-imagename-constants.sh` script for reuse

timebertt committed:
Merge pull request #5054 from rfranzke/fix/grm-constraints

Do not consider GRM webhooks in constraints check

stoyanr committed:
Add unknown status label value (#5041)

rfranzke committed:
`ExtensionValidator` only triggers if spec changed (#5049)

Co-Authored-By: Tim Ebert <tim.ebert@sap.com>

Co-authored-by: Tim Ebert <tim.ebert@sap.com>

rfranzke committed:
Introduce annotation to allow skipping the infrastructure cleanup sleep step (#5044)

* Introduce annotation to allow skipping the infrastructure cleanup sleep step

* Address PR review feedback

rfranzke committed:
Ensure proper extension controller cleanup when seed is deleted (#5047)

Co-Authored-By: Tim Ebert <tim.ebert@sap.com>

Co-authored-by: Tim Ebert <tim.ebert@sap.com>

rfranzke committed:
Activate ReversedVPN feature gate for local development (#5045)

himanshu-kun committed:
Introduce NodeTemplate in Worker extensions API (#4980)

* initial changes

* generated files

* nodeTemplate made optional for backward compatibility

* included suggested refactoring

* changed cloudProfileMachineType to machineType in Values struct

* removed unneeded newline

shafeeqes commented on pkg/operation/botanist/component/nodelocaldns/nodelocaldns.go:
```diff
@@ -0,0 +1,543 @@
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "gardener.kube-system.node-local-dns",
+				Labels: map[string]string{
+					"app": "node-local-dns",
```
You can use `v1beta1constants.LabelApp: LabelValue` and even create a function `getLabels() map[string]string` which returns this exact map, if there are many occurrences of the same.

shafeeqes commented on pkg/operation/botanist/component/nodelocaldns/nodelocaldns.go:
```diff
@@ -0,0 +1,543 @@
+				Labels: map[string]string{
+					"k8s-app":             "node-local-dns",
+					"gardener.cloud/role": "system-component",
+					"origin":              "gardener",
```
You could use
`v1beta1constants.GardenRole:     v1beta1constants.GardenRoleSystemComponent`
`managedresources.LabelKeyOrigin: managedresources.LabelValueGardener`

shafeeqes reviewed (commented):
Just few nits.

ialidzhikov committed:
Update feature gate owners

Signed-off-by: ialidzhikov <i.alidjikov@gmail.com>

stoyanr committed:
Ensure KCM ClusterRoleBinding is not deleted too early (#5062)

ScheererJ committed:
Fixed typo in api-server-proxy configuration file. (#5060)

Now, the access logging should work as desired.

timebertt committed:
Merge pull request #5061 from ialidzhikov/update/feature-gate-owners

Update feature gate owners

himanshu-kun committed:
nodeTemplate added to worker CRD

timebertt committed:
Merge pull request #5067 from himanshu-kun/update-worker-crd

Adds nodeTemplate to worker CRD

acumino committed:
Off

ialidzhikov committed:
Destroy the source BackupEntry only during restore phase (#5068)

Signed-off-by: ialidzhikov <i.alidjikov@gmail.com>

stoyanr committed:
[GEP-17] Set owner check configuration in etcd-main Etcd resource (#4813)

* Set owner check config in etcd-main etcd resource

* Add ownerChecks seed setting

* Address code review comments

ialidzhikov committed:
Change GRM default webhook server port to 10250 (#5076)

Signed-off-by: ialidzhikov <i.alidjikov@gmail.com>

rfranzke committed:
Add seed setting for disabling the `dependency-watchdog` (#5075)

* Documentation

* Add API fields

* `make generate`

* Defaulting

* Adapt seed controller to respect new dwd settings

* Adapt shoot controller to respect new dwd settings

* Address PR review feedback

rfranzke committed:
Eliminate client certificate for `vpa-{admission-controller,recommender,updater}` (#5009)

* Adapt `vpa-*` to use shoot access token

* Rename `enableServiceAccount` to `createServiceAccount`

* Use 'target' in VPA RBAC names when created for shoot

We should probably use 'source' when created for seed, but let's do this when we refactor the Helm chart into a Golang component (otherwise, I would need to cleanup the old `ClusterRole`s in the seed since VPA is not yet managed by GRM, and I don't want to get into this business now with this PR).

rfranzke committed:
Eliminate client certificate for `kube-state-metrics` and `prometheus` (#5008)

* Adapt `kube-state-metrics` to use shoot access token

* Adapt `prometheus` to use shoot access token

For backwards-compatibility, the client certificate cannot yet be removed. This will be done in a future release.

danielfoehrKn committed:
increase infrastructure reconciliation wait timeout (#4881)

ScheererJ committed:
Enable metrics collection of and introduce dashboard for envoy proxy side car used in reversed vpn (#5055)

* Exposed metrics of envoy proxy side car of vpn seed server.

* Import metrics of envoy proxy of reversed vpn to prometheus.

* add dashboard for vpn seed server envoy-proxy sidecar

* update dashboards for envoy proxy

* Fixed legend and titles of reversed vpn envoy proxy dashboards.

* overhaul reversed vpn dashboards and add heatmap

* Removed unnecessary line in reversed vpn scrape config as the default value is used.

* Fixed two more charts with regards minimum value and decimal settings in reversed vpn envoy proxy dashboard.

Co-authored-by: Sebastian Stauch <sebastian.stauch@sap.com>

ScheererJ committed:
Fixed a couple of panels in the istio mesh dashboard and exposed the istio http metrics. (#5077)

Some of the used performance counters were removed from istio quite some time ago.
Apparently, nobody noticed the failing dashboards. Now, at least the dashboards
for virtual services, destination rules and gateways should work again.
Furthermore, a dashboard with a resource that is no longer available was replaced
with workload entries, which we currently do not use.

There were also a couple of panels with incorrect reporter values and one panel
with incorrect tag names.

Last but not least, the additional envoy filters expose the http metrics of istio
so that corresponding dashboards also receive values. The data is currently only
relevant for the reversed-vpn as the apiserver connections go through tcp.

ScheererJ committed:
Add access logging to apiserver-proxy for requests coming from the shoot cluster to kubernetes.default.svc. (#5065)

ialidzhikov committed:
SeedAuthorizer: Allow gardenlet to delete BackupEntry resources (#5066)

* SeedAuthorizer: Allow gardenlet to delete BackupEntry resources

Signed-off-by: ialidzhikov <i.alidjikov@gmail.com>

* Do not send DELETE request when the BackupEntry does not exists

Signed-off-by: ialidzhikov <i.alidjikov@gmail.com>

shafeeqes committed:
Allow to configure `SerializeImagePulls` for kubelet configuration in `Shoot` API (#5074)

* Add `serializeImagePulls` field to `Shoot` API

* Run `make generate`

* Defaulting for the new field

* Populate new fields to `kubelet` component in `OSC` computation

* Apply suggestions from code review

Co-authored-by: Plamen Kokanov <35485709+plkokanov@users.noreply.github.com>

Co-authored-by: Plamen Kokanov <35485709+plkokanov@users.noreply.github.com>

Commit:
Upgrade etcd-druid

from v0.7.0 to v0.7.1

acumino committed:
Add test

rfranzke committed:
Add workaround for projected service account tokens

Workaround https://github.com/kubernetes/kubernetes/issues/82573 - this got fixed with https://github.com/kubernetes/kubernetes/pull/89193 starting with Kubernetes 1.19

ScheererJ committed:
Provide ip address of local node to vpn-seed-server so that the readiness/liveness (#5087)

probes can be removed from the log output.

The readiness/liveness probes clutter the log output. Removing them makes it easier
to reason about what is actually going on.

danielfoehrKn committed:
Merge pull request #5098 from rfranzke/fix/grm-projectedtokenmount-webhook

Add workaround for projected service account tokens

acumino committed:
Add test

acumino committed:
Add test

acumino committed:
Add test

krgostev committed:
[CPM IT]  Adding seed-taint-testrun-value property (#5059)

Commit:
Merge pull request #5037 from abdasgupta/working

Integrate ETCD Druid v0.7.0 to Gardener

Commit:
Merge pull request #5094 from gardener/ci-ixwjldoqu

[ci:component:github.com/gardener/etcd-druid:v0.7.0->v0.7.1]

acumino committed:
Raedy to Review

acumino committed:
Code Rebase

acumino committed:
Testing

acumino committed:
Testing
